src/components/structures/UserMenu.tsx

```diff
@@ -168,28 +174,24 @@ export default class UserMenu extends React.Component<IProps, IState> {
         }
     };
 
-    private addPendingJoinRoom(roomId) {
+    private addPendingJoinRoom(roomId: string): void {
         this.setState({
-            pendingRoomJoin: [
-                ...this.state.pendingRoomJoin,
-                roomId,
-            ],
+            pendingRoomJoin: new Set<string>(this.state.pendingRoomJoin)
+                .add(roomId),
         });
     }
 
-    private removePendingJoinRoom(roomId) {
-        const newPendingRoomJoin = this.state.pendingRoomJoin.filter(pendingJoinRoomId => {
-            return pendingJoinRoomId !== roomId;
-        });
-        if (newPendingRoomJoin.length !== this.state.pendingRoomJoin.length) {
+    private removePendingJoinRoom(roomId: string): void {
+        if (this.state.pendingRoomJoin.has(roomId)) {
+            this.state.pendingRoomJoin.delete(roomId);
             this.setState({
-                pendingRoomJoin: newPendingRoomJoin,
+                pendingRoomJoin: new Set<string>(this.state.pendingRoomJoin),
             })
         }
     }
 
     get hasPendingActions(): boolean {
-        return this.state.pendingRoomJoin.length > 0;
+        return Array.from(this.state.pendingRoomJoin).length > 0;
```

Comment:
Why not `.size`?

Reply:
This still applies, you are turning a Set into an Array twice per render, only to get its `.length`.

Reply:
I missed the `size` property on the Set documentation 😅 
I thought it was a bit convoluted. Thanks for the heads up